### PR TITLE
feat: upsert-d-parameter function

### PR DESCRIPTION
**Pull request by LGLO**
# Description

Uses mint or update d-param transactions.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages.
- [x] New tests are added if needed and existing tests are updated.
~- [ ] Relevant logging and metrics added~
- [x] CI passes. See note on CI.
~- [ ] Any changes are noted in the `changelog.md` for affected crate~
- [x] Self-reviewed the diff

# Note on CI
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges.  Please contact IOG Partner Chains developers to do this
for you.

### CODE REVIEW

File: toolkit/offchain/src/csl.rs [hunk 1]
```diff
@@ -200,6 +201,20 @@ pub(crate) struct TransactionContext {
 }
 
 impl TransactionContext {
+	/// Gets `TransactionContext`, having UTXOs for the given payment key and the network configuration,
+	/// required to perform most of the partner-chains smart contract operations.
+	pub(crate) async fn for_payment_key<C: QueryLedgerState + QueryNetwork>(
```
Comment (AmbientTea) [inline toolkit/offchain/src/csl.rs:206]: Nice

File: toolkit/offchain/src/csl.rs [hunk 2]
```diff
@@ -200,6 +201,20 @@ pub(crate) struct TransactionContext {
...
+		payment_signing_key: [u8; 32],
+		client: &C,
+	) -> Result<TransactionContext, anyhow::Error> {
+		let payment_key = PrivateKey::from_normal_bytes(&payment_signing_key)?;
+		let network = client.shelley_genesis_configuration().await?.network.to_csl();
+		let protocol_parameters = client.query_protocol_parameters().await?;
+		let payment_address = key_hash_address(&payment_key.to_public().hash(), network);
+		let payment_utxos = client.query_utxos(&[payment_address.to_bech32(None)?]).await?;
+		Ok(TransactionContext { payment_key, payment_utxos, network, protocol_parameters })
```
Comment (AmbientTea) [inline toolkit/offchain/src/csl.rs:215]: Maybe `payment_utxos` should just be `utxos`

**Review by AmbientTea — APPROVED** [review]

File: toolkit/offchain/src/d_param.rs [hunk 3]
```diff
@@ -1,13 +1,164 @@
...
+	Ok(McTxHash(res.transaction.id))
+}
+
+fn get_mint_ex_units(response: Vec<OgmiosEvaluateTransactionResponse>) -> anyhow::Result<ExUnits> {
+	// For mint transaction we know there is a single item to evaluate, so matching is trivial.
+	let validator_and_budget = response
+		.first()
+		.ok_or_else(|| anyhow!("Internal error: cannot use evaluateTransaction response"))?;
+	Ok(convert_ex_units(&validator_and_budget.budget))
+}
```
Comment (kpinter-iohk) [inline toolkit/offchain/src/d_param.rs:117]: Can I please have some clarification to why the argument/variable names are different from `get_spend_ex_units`?
`response` vs `costs`, `validator_and_budget` vs `cost`

Comment (LGLO) [reply toolkit/offchain/src/d_param.rs:117]: I can make them equal. For me both are equally good/bad.

Comment (LGLO) [reply toolkit/offchain/src/d_param.rs:117]: I can make them equal. For me both are equally good/bad.

I'll merge it and fix in the PR I have on top of it.

**Pull request merged**